### PR TITLE
Peek definition support for tables, views and stored procedures

### PR DESCRIPTION
**Pull request by sharonravindran**
Add support for peek/go to definition
Add unit tests

Closes Microsoft/vscode-mssql#253 and Microsoft/vscode-mssql#268

### CODE REVIEW

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 1]
```diff
@@ -636,6 +645,131 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+        /// </summary>
+        /// <param name="textDocumentPosition"></param>
+        /// <param name="scriptFile"></param>
+        /// <param name="connInfo"></param>
+        /// <returns> Location with the URI of the script file</returns>
+        internal Location[] GetDefinition(TextDocumentPosition textDocumentPosition, ScriptFile scriptFile, ConnectionInfo connInfo)
+        {
+            Location[] locations = new[] { 
+                    new Location {
+                        Uri = new Uri("C:\\test.txt").AbsoluteUri,
```
Comment (MitchellSternke) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: I'm guessing we shouldn't be using a test URI?

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs [hunk 2]
```diff
@@ -0,0 +1,228 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+#define LIVE_CONNECTION_TESTS
```
Comment (MitchellSternke) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: We define this in one of the build configurations at compile-time, so this probably shouldn't be defined here.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs (hunk 1 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: Good catch! I meant to remove this.

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs (hunk 2 above)
Comment (sharonravindran) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: I'll remove this then.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 3]
```diff
@@ -43,6 +43,8 @@ public sealed class LanguageService
 
         internal const int OnConnectionWaitTimeout = 300000;
 
+        internal const int PeekDefinitionTimeout = 10000;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: >10000 [](start = 51, length = 5)

Minor: Is this 10 seconds? One thing we do in other projects is to define a constant for 1 second and use 10 * Seconds to do this kind of definition. It's handy as it makes clear the unit of measurement. Alternatively a comment also keep this clear.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 4]
```diff
@@ -293,15 +295,22 @@ public void InitializeService(ServiceHost serviceHost, SqlToolsContext context)
             }
         }
 
-// turn off this code until needed (10/28/2016)
-#if false
-        private static async Task HandleDefinitionRequest(
-            TextDocumentPosition textDocumentPosition,
-            RequestContext<Location[]> requestContext)
+        internal static async Task HandleDefinitionRequest(TextDocumentPosition textDocumentPosition, RequestContext<Location[]> requestContext)
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: Shouldn't we skip if intellisense is disabled just like for other request types like CompletionResolveRequest?

Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: Or else if IsQuickInfoEnabled is false? There should be some config flag that skips processing

---
In reply to: [89247609](https://github.com/Microsoft/sqltoolsservice/pull/160#discussion_r89247609) [](ancestors = 89247609)

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 5]
```diff
@@ -636,6 +645,121 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+        {
+            // Parse sql
+            ScriptParseInfo scriptParseInfo = GetScriptParseInfo(textDocumentPosition.TextDocument.Uri);
+            if (RequiresReparse(scriptParseInfo, scriptFile))
+            {
+                ParseAndBind(scriptFile, connInfo);
+            }
+
+            // Return if parse failed
+            if (scriptParseInfo == null)
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: This needs to be checked immediately after GetScriptParseInfo - that's the method that returns null. Otherwise the RequiresReparse check could cause a null ref as it accesses a null info object

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 6]
```diff
@@ -636,6 +645,121 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+            {
+                return null;
+            }
+
+            // Get token from selected text
+            Token token = GetToken(scriptParseInfo, textDocumentPosition.Position.Line + 1, textDocumentPosition.Position.Character);
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: >token [](start = 18, length = 5)

This could use a more descriptive variable name - e.g. What is this token? Is it the schema name?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 7]
```diff
@@ -636,6 +645,121 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+            if (token == null)
+            {
+                return null;
+            }
+            // Strip "[" and "]"(if present) from the token text to enable matching with the suggestions.
+            // The suggestion title does not contain any sql punctuation
+            string tokenText = token.Text.Replace("]",String.Empty).Replace("[",String.Empty);
+
+
+            if (scriptParseInfo.IsConnected && Monitor.TryEnter(scriptParseInfo.BuildingMetadataLock))
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: minor: remove unnecessary whitespace

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 8]
```diff
@@ -636,6 +645,121 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+                try
+                {
+                    // Queue the task with the binding queue    
+                    QueueItem queueItem = this.BindingQueue.QueueBindingOperation(
+                        key: scriptParseInfo.ConnectionKey,
+                        bindingTimeout: LanguageService.PeekDefinitionTimeout,
+                        bindOperation: (bindingContext, cancelToken) =>
+                        {
+                            // Get suggestions for the token
+                            IEnumerable<Declaration> declarationItems;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: >IEnumerable<Declaration> declarationItems; [](start = 28, length = 42)

Please join declaration and assignment here. E.g. IEnumerable<Declaration> declarationItems = Resolver.FindCompletions(...

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+using System;
+using System.IO;
+using System.Collections.Specialized;
+using Microsoft.SqlServer.Management.Smo;
+using Microsoft.SqlTools.ServiceLayer.Connection;
+using Microsoft.SqlTools.ServiceLayer.Workspace.Contracts;
+
+namespace Microsoft.SqlTools.ServiceLayer.LanguageServices
+{
+    public class PeekDefinition
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Please document all public classes and methods

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 10]
```diff
@@ -636,6 +645,121 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+                                            return peekDefinition.GetTableDefinition(tokenText);
+                                        case DeclarationType.View:
+                                            schemaName = this.GetSchemaName(scriptParseInfo, textDocumentPosition.Position, scriptFile);
+                                            return peekDefinition.GetViewDefinition(tokenText, schemaName);
+                                        case DeclarationType.StoredProcedure:
+                                            schemaName = this.GetSchemaName(scriptParseInfo, textDocumentPosition.Position, scriptFile);
+                                            return peekDefinition.GetStoredProcedureDefinition(tokenText, schemaName);
+                                        default:
+                                            return null;
+                                    }
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: This logic should be inside the PeekDefinition class. This service shouldn't know about these implementation details. Given you return null if not found, cleanest thing is move this info the peekDefinition class as-is for now and just call into it.
Separately, we're going to run into scale issues here - there are lots of types in SQL, so we shouldn't rely on switch statement as we'll have 100 entries before long. Options to replace:
- Use a Dictionary<Type,Func<string,string, Location[]> mapping type to a method that supports getting a definition.
- plenty of other patterns, but I'd guess this will work well enough for now. In the long term it might be nice to use dependency injection and have classes that register with supported types so the main logic class never needs to know about this, but it's overkill for now.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+            connectionInfo = connInfo;
+            tempPath = Path.GetTempPath();
+        }
+
+        /// <summary>
+        /// Script a table using SMO and write to a file.
+        /// </summary>
+        /// <param name="tableName">Table name</param>
+        /// <returns>Location object representing URI and range of the script file</returns>
+        internal Location[] GetTableDefinition(string tableName)
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Can't tables have schema names too?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+                                scriptFile.WriteLine(script);
+                            }  
+                        }
+
+                        
+                    }
+                    return GetLocationFromFile(tempFileName, 0);
+                }
+            }
+            return null;
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: The code here is very duplicative for Table, View and Stored Procedure. We're going to have a lot more checks so we can simplify this dramatically. What's different is:
- SMO query to retries the object (table, view, proc)
- script.IndexOf check. The variable we check for is slightly different (basically string.format("CREATE {0}", typeName) should be used)

Based on this, please consider 1 common method that has the main logic, and takes as parameters 1) a getter for the SMO object and 2) The name to query for when trying to find the Create script. in scripts.

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+
+
+        /// <summary>
+        /// Tests the definition event handler. When called with no active connection, no definition is sent
+        /// </summary>
+        [Fact]
+        public void DefinitionsHandlerWithNoConnectionTest()
+        {
+
+            InitializeTestObjects();
```
Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: Minor: minimize unnecessary whitespace

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+            // request the completion list
+            Task handleCompletion = LanguageService.HandleDefinitionRequest(textDocument, requestContext.Object);
+            handleCompletion.Wait(TaskTimeout);
+
+            // verify that send result was not called
+            requestContext.Verify(m => m.SendResult(It.IsAny<Location[]>()), Times.Never());
+
+        }
+
+#if LIVE_CONNECTION_TESTS
```
Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: When do these get run? Do your tests basically require real SMO as it's too hard to mock it out otherwise?

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+        /// <summary>
+        /// Test get definition for a table object with active connection
+        /// </summary>
+        [Fact]
+        public void GetTableDefinitionTest()
+        {
+            ScriptFile scriptFile;
+            ConnectionInfo connInfo = TestObjects.InitLiveConnectionInfo(out scriptFile);
```
Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: What are the script contents? It's not clear what the source of this test is (e.g. source script, backing database) so without this how do I know that "test_table" is present?

**Comment by kevcunnane** [comment]
:clock1:

File: test/Microsoft.SqlTools.ServiceLayer.TestDriver/Tests/LanguageServiceTests.cs [hunk 16]
```diff
@@ -210,6 +210,49 @@ public class LanguageServiceTests
...
+                };
+
+                await testHelper.RequestOpenDocumentNotification(openParams);
+                  
+                Thread.Sleep(500);
+
+                bool connected = await testHelper.Connect(queryTempFile.FilePath, ConnectionTestUtils.LocalhostConnection);
+                Assert.True(connected, "Connection is successful");
+
+                Thread.Sleep(10000);
```
Comment (sharonravindran) [inline test/Microsoft.SqlTools.ServiceLayer.TestDriver/Tests/LanguageServiceTests.cs]: We wait 10 secs for the language service binding. I'll try and optimize this.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs (hunk 5 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: Modified code.

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs (hunk 14 above)
Comment (sharonravindran) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: Yes, these need real SMO to test the scripting and the declarations. These get run when you build and run tests with the integration option. 
Yeah we should be making these required, there are a bunch of tests that are marked live connection.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs (hunk 4 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: Yes, we should add the check here. There are no options for definition per se but I'll add the check for is intellisense enabled

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 17]
```diff
@@ -636,6 +648,102 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+
+            // Get token from selected text
+            Token selectedToken = GetToken(scriptParseInfo, textDocumentPosition.Position.Line + 1, textDocumentPosition.Position.Character);
+            if (selectedToken == null)
+            {
+                return null;
+            }
+            // Strip "[" and "]"(if present) from the token text to enable matching with the suggestions.
+            // The suggestion title does not contain any sql punctuation
+            string tokenText = selectedToken.Text.Replace("]",String.Empty).Replace("[",String.Empty);
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: We need to handle escaped [ ] characters. It's possible to add a table called [Strange\[Name] for example. Please put this in a Utils file since we'll probably need to handle cases like this in other places.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 18]
```diff
@@ -636,6 +650,101 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+        /// <param name="position"></param>
+        /// <param name="scriptFile"></param>
+        /// <returns> schema nama</returns>
+        private string GetSchemaName(ScriptParseInfo scriptParseInfo, Position position, ScriptFile scriptFile)
+        {
+            // Get schema name 
+            string sql = scriptFile.GetLine(position.Line + 1);
+            int startColumn = TextUtilities.PositionOfPrevDelimeter(sql, 0, position.Character);
+            // if no schema name, returns null
+            if( sql[startColumn - 1] != '.')
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: >sql[startColumn - 1] != '.') [](start = 16, length = 28)

We should get the last non-whitespace token instead of assuming the previous character works. The problem is that you can do "SELECT dbo. T1" with a space between the . and the table name. This is  valid and we have to cover these cases. Please also include a test or this scenario :)

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+    /// Peek Definition/ Go to definition implementation
+    /// Script sql objects and write create scripts to file
+    /// </summary>
+    internal class PeekDefinition
+    {
+        private ConnectionInfo connectionInfo;
+        private string tempPath;
+        
+        // Dictionary that holds the script getter for each type
+        private Dictionary<DeclarationType, Func<string, string, StringCollection>> sqlScriptGetters =
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: >Func<string, string, StringCollection>> [](start = 44, length = 39)

Minor: we should probably declare this as a delegate (see [https://msdn.microsoft.com/en-us/library/ms173172.aspx](https://msdn.microsoft.com/en-us/library/ms173172.aspx)) to simplify this and shorten the definition. That would look like 
private delegate StringCollection GetScripts(string schemaName, string objectName). 
Also note that in this pattern, we put schemaName first, then objectName (such as table) as this is the standard we use in other codebases for looking up objects.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+        /// <summary>
+        /// Add getters for each sql object supported by peek definition
+        /// </summary>
+        private void Initialize()
+        {
+            //Add script getters for each sql object
+
+            //Add tables to supported types
+            sqlScriptGetters.Add(DeclarationType.Table, GetTableScripts);
+            sqlObjectTypes.Add(DeclarationType.Table, "Table");
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: When doing the codegen work you'll likely want to make a method that adds to both of these dictionaries in 1 operation. This will make it easier since there's less code and we guarantee that the value is in both dictionaries (which is assumed later on, but right now requires everyone to do the right thing when declaring a new handler)

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+                    : database.Tables[tableName].Script();
+        }
+
+        /// <summary>
+        /// Script a view using SMO
+        /// </summary>
+        /// <param name="viewName">View name</param>
+        /// <param name="schemaName">Schema name </param>
+        /// <returns>String collection of scripts</returns>
+        internal StringCollection GetViewScripts(string viewName, string schemaName)
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: >internal [](start = 8, length = 8)

Can't these be private now? You use the generic method in tests to get these so they no longer need internal accessibility.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+using Microsoft.SqlTools.ServiceLayer.Connection;
+using Microsoft.SqlTools.ServiceLayer.Workspace.Contracts;
+
+namespace Microsoft.SqlTools.ServiceLayer.LanguageServices
+{
+    /// <summary>
+    /// Peek Definition/ Go to definition implementation
+    /// Script sql objects and write create scripts to file
+    /// </summary>
+    internal class PeekDefinition
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Is there a reason this isn't still public?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 23]
```diff
@@ -0,0 +1,210 @@
...
+
+                if (scripts != null)
+                {
+                    int lineNumber = 0;
+                    using (StreamWriter scriptFile = new StreamWriter(File.Open(tempFileName, FileMode.Create, FileAccess.ReadWrite)))
+                    {
+                        
+                        foreach (string script in scripts)
+                        {
+                            if (script.IndexOf(String.Format("CREATE {0}", objectType), StringComparison.OrdinalIgnoreCase) >= 0)
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: >String [](start = 47, length = 6)

please use "string" instead of "String" as that's the code guideline we follow. It's nice if we keep this consistent

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 24]
```diff
@@ -0,0 +1,210 @@
...
+                            {
+                                scriptFile.WriteLine(script);
+                                lineNumber = GetStartOfCreate(script, String.Format("CREATE {0}", objectType));
```
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Again, please use "string" here

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs [hunk 25]
```diff
@@ -0,0 +1,236 @@
...
+            Location[] locations = peekDefinition.GetSqlObjectDefinition(peekDefinition.GetStoredProcedureScripts, objectName, schemaName, objectType);
+            Assert.NotNull(locations);
+            Cleanup(locations);
+        }
+
+        /// <summary>
+        /// Test get definition for a stored procedure object with active connection and no schema
+        /// </summary>
+        [Fact]
+        public void GetStoredProcedureDefinitionWithoutSchemaTest()
```
Comment (kevcunnane) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: Where is the test covering how we handle an unsupported object type? It is important to test negative scenarios and verify we don't crash but instead return empty results cleanly. WIth the new GetSqlObjectDefinition this should be possible? Or is this covered in a different set of tests?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 22 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: It's only expected to be called by the LanguageService so I thought I'll make it internal.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 21 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: I pass these getters to the generic function so they are still internal.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 19 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Modified

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 20 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Modified

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 23 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Modified

File: src/Microsoft.SqlTools.ServiceLayer/SqlContext/SqlToolsSettings.cs [hunk 26]
```diff
@@ -90,6 +90,17 @@ public bool IsQuickInfoEnabled
                     && this.SqlTools.IntelliSense.EnableQuickInfo.Value;
             }
         }
+
+        /// <summary>
+        /// Gets a flag determining if IntelliSense is enabled
+        /// </summary>
+        public bool IsIntellisenseEnabled
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/SqlContext/SqlToolsSettings.cs]: should be IsIntelliSenseEnabled

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 27]
```diff
@@ -0,0 +1,218 @@
...
+        /// <param name="schemaName">Schema name or null</param>
+        /// <param name="objectType">Type of SQL object</param>
+        /// <returns>Location object representing URI and range of the script file</returns>
+        internal Location[] GetSqlObjectDefinition(
+                ScriptGetter sqlScriptGetter, 
+                string objectName, 
+                string schemaName, 
+                string objectType) 
+        {
+            if (this.connectionInfo.SqlConnection != null)
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: why do we null check this here

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 28]
```diff
@@ -0,0 +1,218 @@
...
+
+                if (scripts != null)
+                {
+                    int lineNumber = 0;
+                    using (StreamWriter scriptFile = new StreamWriter(File.Open(tempFileName, FileMode.Create, FileAccess.ReadWrite)))
+                    {
+                        
+                        foreach (string script in scripts)
+                        {
+                            if (script.IndexOf(string.Format("CREATE {0}", objectType), StringComparison.OrdinalIgnoreCase) >= 0)
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: I'd assign string.Format("CREATE {0}", objectType) to a variable to avoid calling same function twice

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 29]
```diff
@@ -0,0 +1,218 @@
...
+        /// <summary>
+        /// Script a table using SMO
+        /// </summary>
+        /// <param name="tableName">Table name</param>
+        /// <param name="schemaName">Schema name</param>
+        /// <returns>String collection of scripts</returns>
+        internal StringCollection GetTableScripts(string tableName, string schemaName)
+        {
+            return (schemaName != null) ? database.Tables[tableName, schemaName]?.Script()
+                    : database.Tables[tableName].Script();
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: does need to be ?. like the other Tables access

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 30]
```diff
@@ -0,0 +1,218 @@
...
+        /// </summary>
+        /// <param name="declarationItems"></param>
+        /// <param name="tokenText"></param>
+        /// <param name="schemaName"></param>
+        /// <returns>Location object of the script file</returns>
+        internal Location[] GetScript(IEnumerable<Declaration> declarationItems, string tokenText, string schemaName)
+        {
+            foreach (Declaration declarationItem in declarationItems)
+            {
+                if (declarationItem.Title.Equals(tokenText))
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Title can't be null?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 31]
```diff
@@ -0,0 +1,218 @@
...
+        internal delegate StringCollection ScriptGetter(string objectName, string schemaName);
+        
+        // Dictionary that holds the script getter for each type
+        private Dictionary<DeclarationType, ScriptGetter> sqlScriptGetters =
+            new Dictionary<DeclarationType, ScriptGetter>();
+
+        //Dictionary that holds the object name (as appears on the TSQL create statement)
+        private Dictionary<DeclarationType, string> sqlObjectTypes = new Dictionary<DeclarationType, string>();
+
+        private Database database
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: property names should be Capitalized..so **private Database Database** would be a better definition here.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 32]
```diff
@@ -0,0 +1,218 @@
...
+            new Dictionary<DeclarationType, ScriptGetter>();
+
+        //Dictionary that holds the object name (as appears on the TSQL create statement)
+        private Dictionary<DeclarationType, string> sqlObjectTypes = new Dictionary<DeclarationType, string>();
+
+        private Database database 
+        {
+            get
+            {
+                Server server = new Server(this.connectionInfo.SqlConnection.DataSource);
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: should we null check this.connectionInfo.SqlConnection here?

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 33]
```diff
@@ -0,0 +1,218 @@
...
+                return server.Databases[this.connectionInfo.SqlConnection.Database];
+            }
+        }
+        
+        internal PeekDefinition(ConnectionInfo connInfo)
+        {
+            connectionInfo = connInfo;
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: let's use a "this." prefix when access instance fields and properties.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs [hunk 34]
```diff
@@ -0,0 +1,218 @@
...
+            DirectoryInfo tempScriptDirectory = Directory.CreateDirectory( Path.GetTempPath()+ "mssql_definition");
```
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: nit: spacing..should be **Directory.CreateDirectory(Path.GetTempPath() + "mssql_definition");**

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 22 above)
Comment (kevcunnane) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: OK sounds good

---
In reply to: [90150351](https://github.com/Microsoft/sqltoolsservice/pull/160#discussion_r90150351) [](ancestors = 90150351)

**Comment by erickangMSFT** [comment]
Please try Go To Definition and Peek Definition with the attached database.

1. Run 1 setup.sql to create ProductCatalog database
2. Run setup-temporal.sql to populate data
3. Using temporal.sql connected to ProductCatalog database, try the features.

It seems IntelliSense suggestion list and Go To Definition and Peek Definition do not function as expected.
The ProductCatalog database is a sample demo from https://github.com/Microsoft/sql-server-sample 
[4 temporal.zip](https://github.com/Microsoft/sqltoolsservice/files/623550/4.temporal.zip)

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 32 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: Yes. I should have move the check up when I refactored the code. Modifying this.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs [hunk 35]
```diff
@@ -636,6 +650,110 @@ internal CompletionItem ResolveCompletionItem(CompletionItem completionItem)
...
+            string tokenText = TextUtilities.RemoveSquareBracketSyntax(selectedToken.Text);
+
+            if (scriptParseInfo.IsConnected && Monitor.TryEnter(scriptParseInfo.BuildingMetadataLock))
+            {
+                try
+                {
+                    // Queue the task with the binding queue    
+                    QueueItem queueItem = this.BindingQueue.QueueBindingOperation(
+                        key: scriptParseInfo.ConnectionKey,
+                        bindingTimeout: LanguageService.PeekDefinitionTimeout,
```
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: Nothing. The users sees the peek window on return. I've set it to ten seconds to give it enough time to find resolutions(worst case).

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs (hunk 4 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: I think if IntelliSense is disabled in general then PeekDefinition should be disabled too but the menu option that doesn't do anything is not good. I'll ask Eric what he wants.

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs (hunk 27 above)
Comment (sharonravindran) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/PeekDefinition.cs]: This check is here since the database get doesn't check for null. I'll refactor the checks.

File: test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs (hunk 25 above)
Comment (sharonravindran) [inline test/Microsoft.SqlTools.ServiceLayer.Test/LanguageServer/PeekDefinitionTests.cs]: I've added the test for the unsupported scenario

File: src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs (hunk 35 above)
Comment (kburtram) [inline src/Microsoft.SqlTools.ServiceLayer/LanguageServices/LanguageService.cs]: this will block Intellisense during this time which might be not the best UX.  We can start with this and see how it works.
